tqdm to logging

de-code opened the pull request:
see #313
implementation of https://github.com/tqdm/tqdm/issues/313#issuecomment-812224667

/cc @casperdcl

de-code committed:
added logging_tqdm

fixed comma in type hint

added logging_tqdm

added documentation to class

added documentation to README

python 2

flake8

also test with default logger

isort

fixed readme title underline too short

improved coverage

brtasavpatel commented:
@de-code Progress bar update shows up in the new line. Is this expected?

![image](https://user-images.githubusercontent.com/38225409/120205646-3169ce80-c1f8-11eb-9413-f102720e2624.png)

de-code commented:
> @de-code Progress bar update shows up in the new line. Is this expected?

Yes, that is intentional (except for maybe the last line, not sure why it is logging the same progress twice - could add a check for that). It is passing the progress to the regular logging module. No magic there. On the plus side, in a log file it will also still look sensible.
You might be looking for a different functionality.

brtasavpatel commented:
> > @de-code Progress bar update shows up in the new line. Is this expected?
> 
> Yes, that is intentional (except for maybe the last line, not sure why it is logging the same progress twice - could add a check for that). It is passing the progress to the regular logging module. No magic there. On the plus side, in a log file it will also still look sensible.
> You might be looking for a different functionality.

I am looking for a solution where my progress bar won't look weird in kubernetes logs, at the same time I am my progress bar is usually for a million files or so that means this will add lot of lines in my logs.

de-code commented:
> I am looking for a solution where my progress bar won't look weird in kubernetes logs, at the same time I am my progress bar is usually for a million files or so that means this will add lot of lines in my logs.

I think this is where the `mininterval` comes in handy (core feature). With the `logging_tqdm` in this PR, the default is set to `1` (i.e. 1 second). Which I found personally useful, but you could increase that further. Therefore it doesn't matter how many rows you process but how frequent you want to see a new log message.

What I haven't tested is whether the `mininterval` is respected when updating the progress description (if that is what you are doing).

casperdcl commented:
I think @brtasavpatel is referring to duplicating the output: once (normally) to `stdout` as well as to a log file

de-code committed:
avoid logging same message twice

de-code committed:
added mininterval to example

de-code commented:
I've updated the PR to avoid logging the same message twice (which I think may have happened at the end).

Regarding the logging, I think the confusion came from the sleep in https://github.com/tqdm/tqdm/pull/1172#issuecomment-851516343 being the same as (or larger than) `mininterval`, 1 second. Which results in a message for about every item. That gives the impression that a message is always logged for every item. But if say `mininterval` was set to one minute, it would only log once every minute (independent of how many items it processed within that period).

In my example I had the sleep set to `0.3`. I have now added a comment to the example and also explicitly include `mininterval`. Maybe the comment could be expanded further.

```python
import logging
from time import sleep
from tqdm.contrib.logging import logging_tqdm

LOG = logging.getLogger(__name__)

if __name__ == '__main__':
    logging.basicConfig(level=logging.INFO)
    with logging_tqdm(range(10), mininterval=1):
        sleep(0.3)  # assume processing one item takes less than mininterval
```

de-code committed:
fixed self passed twice to super logging_tqdm __init__

de-code committed:
added logger to example

de-code committed:
fixed example using tqdm incorrectly

de-code committed:
removed example comment regarding logging restored

de-code committed:
more thorough iterable test

de-code committed:
raise explicit error with multiple positional args

de-code commented:
Would it be worth having something like a `tqdm.auto` that would select `logging_tqdm` if stderr doesn't seem to be tty?

de-code committed:
removed commented out test code

de-code commented:
Is this still desired?

brtasavpatel commented:
I would definitely like this to be added in

de-code committed:
Merge branch 'master' into tqdm-to-logging

de-code commented:
Okay, I merged with master.
Let me know if anything is missing.

tesfaldet commented:
Is this still being worked on? I'd love to have this feature as well for use with logging in [Hydra](https://hydra.cc/docs/tutorials/basic/running_your_app/logging/) (which uses python's logging stdlib) when I need to check the log files from experiment sweeps.

de-code commented:
Hi @casperdcl just wondering anything should be done from my side in order to get this merged?

de-code commented:
Oh dear, I missed the anniversary of this PR. :birthday: 

Hi @casperdcl it seems some people are interested in getting this merged.
It would be good to hear back either way.

stdavis commented:
I used this code in my project and it works great! I would love to see this merged!
